Configure project to fetch Gutenberg as an AAR from Bintray by default

mokagio opened the pull request:
☝️ Before merging this we need to be publishing Gutenberg to the WordPress org. For that, we need to finish the work in https://github.com/WordPress/gutenberg/pull/27836.

---

This makes the project use the Gutenberg React Native Android bridge AAR version published to [Bintray](http://bintray.com/).

To test, make sure to set `wp.BUILD_GUTENBERG_FROM_SOURCE=false` in your `gradle.properties` to make the build system use the Bintray package, then run the app and try to publish a post. Verify that the build and editor still work with the property set to `true`.

## Current limitations

- The package used comes from my Bintray repository. This needs to be addressed before merging this PR, but before doing so it would be appropriate to be able to run the Bintray publishing from CI, which is not possible yet.
- There is no support for composite build, yet. That is, we still rely on having `mobile-gutenberg` and `gutenberg` as submodules in the repository. This will likely be addressed in a dedicated followup PR, given the scope of the task.

---

PR submission checklist:

- [ ] ~~I have considered adding unit tests where possible.~~
- [ ] ~~I have considered adding accessibility improvements for my changes.~~
- [ ] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~

mokagio committed:
Extract "build Gutenberg from source" in WP editor for future reuse

mokagio committed:
Configure project to fetch Gutenberg as an AAR from Bintray by default

mokagio commented:
Asking @mchowning and @loremattei for review, despite the PR still being a draft to double-check everything works and makes sense. Thanks! 🙇

mchowning commented on WordPress/build.gradle:
```diff
@@ -331,6 +339,22 @@ dependencies {
+    //
+    // A starting point might be the fact that the
+    // @wordpress_react-native-bridge-release AAR contains only the
+    // -release.aar version of Hermes in its libs folder
```
In testing this earlier, I also noticed that the hermes dependency bundled in the aar was way too small of a file (only 22 bytes). 😕

mchowning commented:
I only had time to do a quick test, but using 959b3c5caa from this PR and 39bb2901be from [the gutenberg PR](https://github.com/WordPress/gutenberg/pull/27836), both the from-source and from-binary builds worked for me! 🎉 This is some great work @mokagio !

oguzkocer commented on settings.gradle:
```diff
@@ -56,6 +58,7 @@ if (properties.getOrDefault('wp.BUILD_GUTENBERG_FROM_SOURCE', false).toBoolean()
 
 gradle.ext.fluxCBinaryPath = "com.github.wordpress-mobile.WordPress-FluxC-Android"
 gradle.ext.wputilsBinaryPath = "org.wordpress:utils"
+gradle.ext.gutenbergMobileBinaryPath = "com.github.wordpress-mobile.gutenberg-mobile:react-native-gutenberg-bridge"
```
I don't think we need to put this property in the `settings.gradle` file since we don't need to use it in the `settings.gradle` itself. I suggest moving this to the root `build.gradle` file's properties, maybe above/below the new `gutenbergMobileVersion` property. What do you think?

oguzkocer replied:
@mokagio Please ignore this comment. My memory is clearly failing me since we need this property for composite builds 🤦

oguzkocer reviewed (commented):
@mokagio This looks good to me for the Phase 1 merge. I left a minor comment that's not a blocker (I can make the change as part of the composite build work if you prefer) and it's just waiting the base branch to be updated. There also seems to be a second similar PR, so I wasn't sure which one you wanted to merge. If you can set things up for merge, I can do that tomorrow.

oguzkocer commented on WordPress/build.gradle:
```diff
@@ -186,6 +186,14 @@ androidExtensions {
 }
 
 dependencies {
+    // Gutenberg Mobile
+    //
+    // Use a binary from Bintray by default, but can optionally be configured
+    // to use a local build via local-builds.gradle
+    implementation "$gradle.ext.gutenbergMobileBinaryPath:$gutenbergMobileVersion"
+    testImplementation "$gradle.ext.gutenbergMobileBinaryPath:$gutenbergMobileVersion"
+    debugImplementation "$gradle.ext.gutenbergMobileBinaryPath:$gutenbergMobileVersion"
+
```
```suggestion
```

I don't think we need this dependency in the WordPress module. As far as I know Gutenberg is only enabled through the `WordPressEditor` module, so we should be safe to remove this.

I've tested this change and the editor worked as expected for multiple different blocks.

oguzkocer commented on WordPress/build.gradle:
```diff
@@ -331,6 +339,22 @@ dependencies {
     implementation (project(path:':libs:editor:WordPressEditor')) {
         exclude group: 'org.wordpress', module: 'utils' 
     }
+
+    // Hermes is a React Native dependency that we somehow lost at runtime
+    // since moving to the Bintray binary for Gutenberg.
+    //
+    // These cannot be defined in the WordPressEditor project itself, or we'll
+    // get a "Direct local .aar file dependencies are not supported when
+    // building an AAR." error.
+    //
+    // FIXME: In the long run, we'll want to read this from the Gutenberg AAR.
+    //
+    // A starting point might be the fact that the
+    // @wordpress_react-native-bridge-release AAR contains only the
+    // -release.aar version of Hermes in its libs folder
+    debugRuntimeOnly files('./../libs/hermes/hermes-debug.aar')
+    releaseRuntimeOnly files('./../libs/hermes/hermes-release.aar')
+
```
I think this is supposed to be a part of the `WordPressEditor` module as well. WordPress module shouldn't need the hermes binary.

I know this is just a temporary hack 🤞 but I think it'd be good to keep things in their proper place so as to not confuse ourselves.

mokagio replied:
>  I think it'd be good to keep things in their proper place so as to not confuse ourselves.

💯 thank you for pointing this out. I'll have a look.

mokagio replied:
Hey @oguzkocer, I had forgotten about my comment regarding the "Direct local .aar file dependencies are not supported when building an AAR" error" and I tried moving the these statements in the `build.gradle` for the editor. I got that error again.

As far as I can understand, we can't have an AAR that includes another AAR.

This is the full message:

> Direct local .aar file dependencies are not supported when building an AAR. The resulting AAR would be broken because the classes and Android resources from any local .aar file dependencies would not be packaged in the resulting AAR. Previous versions of the Android Gradle Plugin produce broken AARs in this case too (despite not throwing this error).

This SO answer [suggests](https://stackoverflow.com/a/63665094/809944) to import the second AAR (in our case the Hermes one) from a repository, but I think we should be able to find a way to make the `gutenberg` code package it for us properly (see https://github.com/orgs/wordpress-mobile/projects/80#card-52385821).

I could address part of the "keep things in their proper place so as to not confuse ourselves" by moving the lib in the editor subfolder, but the includes would still be here. What do you think? Or have you got another suggestion on how to get this to work? 🙏

oguzkocer replied:
[This](https://github.com/wordpress-mobile/WordPress-Android/pull/13014/files#diff-46e543167c983e22ee4eed026dfc1ad6624f9878ef3b7826acea65b402811b97R97-R99) is the way we have done it before, as a temporary measure of course. Does that answer your question?

mokagio replied:
Thank you @oguzkocer. I did try that approach originally and it resulted in that "Direct local .aar file dependencies are not supported when building an AAR..." error.

I updated it to what's in f2cf853 and it works when running from `./gradlew` and from Android Studio both.

Hopefully it won't be a long lived measure anyway, but I'm glad you helped me make it better.

oguzkocer reviewed (changes requested):
@mokagio I am so sorry to leave a second review, but as I was working on composite build, I've noticed a couple issues and thought it'd be good to address them before we merge this in. I hope that's OK!

mokagio commented:
@oguzkocer thank you for the reviews(s 😄) and for helping me out shaping this to be in a better form. 

I'll try to address your comments ASAP 👍

mokagio committed:
Update gutenberg binary version

mokagio committed:
Remove unnecessary Gutenberg import from WordPress/build.gradle

We only use it in the editor library, so the import that's there is
enough.

mokagio committed:
Move Hermes library to libs/editor/WordPressEditor/libs

That's where it should live because it's required only by
WordPressEditor, even though at this point in time we need to import it
via the WordPress because of the "Direct local .aar dependencies are not
supported" error.

mokagio committed:
Use a parametric path for the Hermes library include

mokagio committed:
Move Hermes import to WordPressEditor

oguzkocer reviewed (approved):
For phase 1, this looks good to me. I think we should mark this PR as ready for review and merge it directly to `gutenberg-composite-build` branch. However, before that, one thing we should probably track is the commit we use to build the bintray version of gutenberg for each PR. Since the `gutenbergMobileVersion` does not contain that information, we should probably add the gutenberg and gb-mobile commits as a comment above the `ext.gutenbergMobileVersion` line, so we can reproduce a build if necessary.

What do you think @mokagio?

mokagio commented:
> one thing we should probably track is the commit we use to build the bintray version of gutenberg for each PR

Absolutely! I hadn't considered that and simply rolled with the approach I originally used to generate the version based on a timestamp. That was useful at the start, but we should definitely use the commit instead.

In fact, rather than compromise and leave a comment which could easily become outdated, I'll just go ahead and generate a new version that uses the SHA as its identifier.

I'll use the commit SHA from `gutenberg-mobile` repo, as checking out that one will also checkout the appropriate `gutenberg` commit.

In WordPress-Utils-Android, we [use `<PR number>-<commit SHA>` as the version](https://github.com/wordpress-mobile/WordPress-Utils-Android/blob/ae5cb8e20e901856f4c28e22e60ffddee0f501a8/.circleci/config.yml#L50-L70). We might apply the same approach here once I'll get CI publishing to work.

mokagio committed:
Switch to React Native bridge version identified by a SHA

This will make it easier to reconstruct the state of the codebase if we
want to run a local build using the code that produced that binary.

mokagio commented:
@oguzkocer, I thought f2cf853 solved the "Direct local `.aar`..." issue, but I discovered it doesn't 😞 

If I run `assembleRelease` from there, I get that error.

Something worth noting is that `assembleJalapenoDebug` (which is what CI uses) doesn't incur in that problem, and neither does `assembleVanillaRelease`. 🤔 Am I just holding it wrong? Should I not use `assembleRelease`?

If the answer to that question is yes, then I guess the PR is really, finally ready to merge 😄

oguzkocer commented:
@mokagio That `direct local .aar` issue  is related to us trying to embed Hermes binary in the Gutenberg end, right? Since that embed is not working right now and we have a workaround for it (where we add it manually in WPAndroid) shouldn't we just remove that? It has given me some issues on my end as well and I just commented it out.

I also don't think `assembleRelease` is a valid command simply because it's lacking a variant such as `vanilla`, `jalapeno`, `wasabi` etc. However, maybe it's supposed to select the default variant, I am not sure.

I am not able to give this a try today, but I can try it out tomorrow if you can't do it.

mokagio commented:
> Since that embed is not working right now and we have a workaround for it (where we add it manually in WPAndroid) shouldn't we just remove that? It has given me some issues on my end as well and I just commented it out.

😮 yeah that makes sense. I tried it and it worked 👏

mokagio committed:
Point gutenberg-mobile to commit with Gutenberg building Bintray AAR

mchowning commented:
> I also don't think assembleRelease is a valid command simply because it's lacking a variant such as vanilla, jalapeno, wasabi etc. However, maybe it's supposed to select the default variant, I am not sure.

FWIW, I think `assembleRelese` _should_ run _all_ the variants:

```
$ ./gradlew assembleRelease --dry-run | grep -E 'assemble.*Release'
:@react-native-community_masked-view:assembleRelease SKIPPED
:@react-native-community_slider:assembleRelease SKIPPED
...
:react-native-gutenberg-bridge:assembleRelease SKIPPED
...
:WordPress:assembleJalapenoRelease SKIPPED
:WordPress:assembleVanillaRelease SKIPPED
:WordPress:assembleWasabiRelease SKIPPED
:WordPress:assembleZalphaRelease SKIPPED
:WordPress:assembleRelease SKIPPED
:WordPressMocks:assembleRelease SKIPPED
...
```

mokagio committed:
Switch from @mokagio's to @wordpress-mobile's Bintray org

We were on mine while ironing out the details of the setup. Now that it
works fine, I don't fear "polluting" the official org as much.

mokagio committed:
Update gutenberg-mobile with latest Gutenberg building Bintray AAR

mokagio committed:
Merge branch 'gutenberg-composite-build' into react-native-from-bintray-binary-only

mokagio commented:
@oguzkocer @mchowning gentlemen, I think this is now really ready to merge. I tried it myself but since there's a lot of stuff going on under the hood that I did to address some of the comments since @oguzkocer approval, can I bug you to have another look to see if this works for you, too?

(If it does, feel free to merge)

Thanks a lot 🙇

oguzkocer reviewed (approved):
This looks good to me as a first step, so I'd vote for merging it. The only issue seems to be `ci/circleci: gutenberg-bundle-build` CI task. If that's not a blocker (something we can address in another PR or just ignore), feel free to merge this to the WIP branch @mokagio.

Please note that I'll be making a lot of changes to these for composite builds, so it helps me to merge this as soon as possible.

mokagio commented:
Thanks @oguzkocer 👍 

Interesting that the build failure is due to CI running out of memory:

> Loading dependency graph, done.
> error ENOMEM: not enough memory, write. Run CLI with --verbose flag for more details.

That's something I'm seeing on the `gutenberg-mobile` PR to publish to Bintray on CI, too. Maybe some of the work I did broke something in the Node setup, or we need to better tune the memory config in CI, or both.

mchowning commented:
> Interesting that the build failure is due to CI running out of memory:

> > Loading dependency graph, done.
error ENOMEM: not enough memory, write. Run CLI with --verbose flag for more details.

> That's something I'm seeing on the gutenberg-mobile PR to publish to Bintray on CI, too. Maybe some of the work I did broke something in the Node setup, or we need to better tune the memory config in CI, or both.

We've been seeing that error sometimes on WPAndroid PRs, so most likely it's a preexisting issue (i.e., we need to better tune the memory config on CI) and not a new one (most WPAndroid devs probably never see that error because CI is pulling a previously built bundle from the cache for them).

mokagio commented:
Thanks for the extra context @mchowning.

I've seen that often while working on https://github.com/wordpress-mobile/gutenberg-mobile/pull/2945 and I think I got it to a stable place memory wise 🤞
